### DISCUSSION TITLE
arguing the security properties of xsnap (XS, SES, JS)

### DISCUSSION THREAD

**Discussion by dckc** [Agoric/agoric-sdk#7858] [Ideas]
@erights , you asked me how I'd go about this... the question has been rattling around in my head and it's starting to take shape... let me know if I should move this under the Agoric org. during this early brainstorming stage.
cc @warner @dtribble @kriskowal

  - [ ] conjecture: XS / xsnap is memory-safe
    - [ ] define memory-safety
    - [ ] properties of the [XS machine](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/xs/XS%20in%20C.md#machine): "A machine uses chunks to store strings, bytecodes, array buffers, big int values, and others. ... A machine uses a heap and a stack of slots."
    - [ ] properties of the bytecode interpreter
    - [ ] burden of proof: host functions for string manipulation, etc.
    - [x] empirical measurements of xsnap: LOC, byte size of .o files and ELF executable, library dependencies
    - [ ] snapshot / restore correctness
       - [ ] define snapshot / restore (defensive?) correctness
       - [ ] conjecture: xsnap.c is correct
    - [ ] Agoric additions to xsnap.c:
        - [ ] netstrings
        - [ ] I/O loop
           - [x] review by moddable? Did we use `xsBeginHost` and such correctly?
  - [ ] conjecture: SES is an ocap system / language
     - [ ] formalize SES security properties: "For example, Politz et al. [Pol+11a] present a type system for λJS which captures the informal restrictions enforced by ADSafe." -- [Certified semantics and analysis of JavaScript](https://tel.archives-ouvertes.fr/tel-01478722/document) p. 32 section 2.2.2
  - [ ] conjecture: XS correctly implements JS / SES
     - [ ] conjecture: XS conformance issues do not represent defensive correctness risks

dump of open tabs:

 - [javascript formal semantics - Google Search](https://www.google.com/search?sxsrf=ALeKk01uCGr1dJ88R2q_WnWeX58UbNts_w%3A1611152528161&ei=kDwIYIqrCca6tQbAnp_4AQ&q=javascript+formal+semantics&oq=javascript+formal+semantics&gs_lcp=CgZwc3ktYWIQAzIJCAAQyQMQFhAeOgoIABCRAhBGEPkBOgUIABCRAjoICC4QsQMQgwE6CwguELEDEMcBEKMCOggIABCxAxCDAToICC4QxwEQowI6BAgAEEM6CAgAEMkDEJECOgcIABCxAxBDOgcIABDJAxBDOgwIABDJAxBDEEYQ-QE6CAgAELEDEMkDOgUIABCxAzoCCAA6BQgAEMkDOgYIABAWEB46BwgAEMkDEA06BAgAEA1Q-KkGWIjKBmD_ygZoAHABeACAAXyIAe4QkgEEMjUuMpgBAKABAaoBB2d3cy13aXrAAQE&sclient=psy-ab&ved=0ahUKEwiKs_rw2qruAhVGXc0KHUDPBx8Q4dUDCA0&uact=5)
 - [Memory safety - Wikipedia](https://en.wikipedia.org/wiki/Memory_safety)
 - [What is memory safety? - The PL Enthusiast](http://www.pl-enthusiast.net/2014/07/21/memory-safety/)
 - [The Meaning of Memory Safety](https://nebelwelt.net/teaching/17-527-SoftSec/slides/02-memory_safety.pdf)
   - "Capability-based memory safety is a form of type safety with two
types: pointer-types and scalars." -- Prof. [Mathias Payer](https://nebelwelt.net/) Spring 2017
   - [CS527 Software Security](https://nebelwelt.net/teaching/17-527-SoftSec/)
 - [What is Memory Safety in Programming? | by Arun Rajeevan | Medium](https://arunrajeevan.medium.com/what-is-memory-safety-in-programming-1e5300fe6c48)
 - [KJS: A Complete Formal Semantics of JavaScript - FSL](http://fsl.cs.illinois.edu/index.php/KJS:_A_Complete_Formal_Semantics_of_JavaScript)
 - [javascript-semantics/README.md at master · kframework/javascript-semantics](https://github.com/kframework/javascript-semantics/blob/master/README.md)
 - [Certified semantics and analysis of JavaScript](https://tel.archives-ouvertes.fr/tel-01478722/document)
 - [An Operational Semantics for JavaScript](https://theory.stanford.edu/people/jcm/papers/aplas08-camera-ready.pdf)
 - [A Structural Operational Semantics for JavaScript](https://theory.stanford.edu/~ataly/Talks/aplas08.pdf)

**Comment by dckc** [comment]
## Hardware support? MPX in Intel Skylake


From [What is memory safety? \- The PL Enthusiast](http://www.pl-enthusiast.net/2014/07/21/memory-safety/):
> ## Pointers as capabilities
> 
> The definition I prefer, which I’ll now present, is inspired by Santosh Nagarakatte‘s work (done with several collaborators) on SoftBound, an approach for enforcing memory safety in C programs (which has inspired [Intel hardware support][i1], coming soon).

[Intel MPX \- Wikipedia](https://en.wikipedia.org/wiki/Intel_MPX) citing "Intel Software Development Emulator". Intel. 2012-06-15:
> Intel MPX was introduced as part of the Skylake microarchitecture.

`grep -i mpx /proc/cpuinfo` shows `mpx`, suggesting it's on my machine.

[Intel® Memory Protection Extensions (Intel® MPX) support in the GNU toolchain][i1] says it's deprecated, and Wikipedia goes on to say "In practice, there have been too many flaws discovered in the design for it to be useful, and support has been deprecated or removed from most compilers and operating systems." But it still looks interesting, to me. Wikipedia doesn't note the connection to capabilities. I wonder how similar it is to CHERI.

[i1]: https://software.intel.com/content/www/us/en/develop/blogs/intel-memory-protection-extensions-intel-mpx-support-in-the-gnu-toolchain.html

**Comment by dckc** [comment]
## Executable: 1.5M

2.7M for debug

```
agoric-sdk/packages/xsnap$ ls -lsh build/bin/lin/*/xsnap
2.7M -rwxrwxr-x 1 connolly connolly 2.7M Jan 23 20:03 build/bin/lin/debug/xsnap
1.5M -rwxrwxr-x 1 connolly connolly 1.5M Jan 23 20:03 build/bin/lin/release/xsnap
```

## Compiled C: 2,084KB

```
agoric-sdk/packages/xsnap/build/tmp/lin/release/xsnap$ ls -sh *.o
8.0K xsAll.o         96K xsDataView.o    36K xsJSON.o       36K xsObject.o      28K xsScope.o
 60K xsAPI.o        120K xsDate.o        60K xsLexical.o    12K xsPlatforms.o   16K xsScript.o
8.0K xsArguments.o  8.0K xsDebug.o       48K xsMapSet.o    4.0K xsProfile.o     68K xsSnapshot.o
 80K xsArray.o       16K xsDefaults.o    20K xsMarshall.o   32K xsPromise.o    8.0K xsSourceMap.o
 60K xsAtomics.o     44K xsdtoa.o        24K xsMath.o       12K xsProperty.o    52K xsString.o
 64K xsBigInt.o      12K xsError.o       68K xsMemory.o     36K xsProxy.o       16K xsSymbol.o
4.0K xsBoolean.o     24K xsFunction.o    36K xsModule.o     32K xsRegExp.o     108K xsSyntaxical.o
 96K xsCode.o        24K xsGenerator.o   48K xsnap.o       292K xsre.o          60K xsTree.o
 48K xsCommon.o      24K xsGlobal.o      16K xsNumber.o     88K xsRun.o         32K xsType.o


agoric-sdk/packages/xsnap/build/tmp/lin/release/xsnap$ ls -s *.o >,siz
agoric-sdk/packages/xsnap/build/tmp/lin/release/xsnap$ python

Python 2.7.17 (default, Sep 30 2020, 13:38:04) 
>>> lines = open(',siz').readlines()
>>> sum([int(line.strip().split(None, 1)[0]) for line in lines if line.strip()])
2084
```

## Dependencies: libc, libpthread only

The debug build also has some socket stuff but we no longer use that in production: https://github.com/Agoric/agoric-sdk/issues/2216

```
agoric-sdk/packages/xsnap$ ldd build/bin/lin/release/xsnap 
	linux-vdso.so.1 (0x00007ffe61391000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f57743c6000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f57741a7000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f5773db6000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f5774ae5000)

agoric-sdk/packages/xsnap$ nm build/bin/lin/release/xsnap | grep 'U '
                 U acos@@GLIBC_2.2.5
                 U acosh@@GLIBC_2.2.5
                 U asin@@GLIBC_2.2.5
                 U asinh@@GLIBC_2.2.5
                 U atan2@@GLIBC_2.2.5
                 U atan@@GLIBC_2.2.5
                 U atanh@@GLIBC_2.2.5
                 U calloc@@GLIBC_2.2.5
                 U cbrt@@GLIBC_2.2.5
                 U cos@@GLIBC_2.2.5
                 U cosh@@GLIBC_2.2.5
                 U __ctype_tolower_loc@@GLIBC_2.3
                 U __errno_location@@GLIBC_2.2.5
                 U exit@@GLIBC_2.2.5
                 U exp@@GLIBC_2.2.5
                 U expm1@@GLIBC_2.2.5
                 U fclose@@GLIBC_2.2.5
                 U fdopen@@GLIBC_2.2.5
                 U feof@@GLIBC_2.2.5
                 U fflush@@GLIBC_2.2.5
                 U fgetc@@GLIBC_2.2.5
                 U floor@@GLIBC_2.2.5
                 U fmod@@GLIBC_2.2.5
                 U fopen@@GLIBC_2.2.5
                 U __fprintf_chk@@GLIBC_2.3.4
                 U fputc@@GLIBC_2.2.5
                 U fputs@@GLIBC_2.2.5
                 U fread@@GLIBC_2.2.5
                 U free@@GLIBC_2.2.5
                 U fwrite@@GLIBC_2.2.5
                 U gettimeofday@@GLIBC_2.2.5
                 U hypot@@GLIBC_2.2.5
                 U __isoc99_fscanf@@GLIBC_2.7
                 U __libc_start_main@@GLIBC_2.2.5
                 U localtime@@GLIBC_2.2.5
                 U log10@@GLIBC_2.2.5
                 U log1p@@GLIBC_2.2.5
                 U log2@@GLIBC_2.2.5
                 U log@@GLIBC_2.2.5
                 U __longjmp_chk@@GLIBC_2.11
                 U malloc@@GLIBC_2.2.5
                 U memcmp@@GLIBC_2.2.5
                 U __memcpy_chk@@GLIBC_2.3.4
                 U memcpy@@GLIBC_2.14
                 U memmove@@GLIBC_2.2.5
                 U memset@@GLIBC_2.2.5
                 U mktime@@GLIBC_2.2.5
                 U nearbyint@@GLIBC_2.2.5
                 U pow@@GLIBC_2.2.5
                 U __printf_chk@@GLIBC_2.3.4
                 U pthread_cond_destroy@@GLIBC_2.3.2
                 U pthread_cond_init@@GLIBC_2.3.2
                 U pthread_cond_signal@@GLIBC_2.3.2
                 U pthread_cond_timedwait@@GLIBC_2.3.2
                 U pthread_cond_wait@@GLIBC_2.3.2
                 U pthread_mutex_destroy@@GLIBC_2.2.5
                 U pthread_mutex_init@@GLIBC_2.2.5
                 U pthread_mutex_lock@@GLIBC_2.2.5
                 U pthread_mutex_unlock@@GLIBC_2.2.5
                 U pthread_self@@GLIBC_2.2.5
                 U puts@@GLIBC_2.2.5
                 U qsort@@GLIBC_2.2.5
                 U rand@@GLIBC_2.2.5
                 U realloc@@GLIBC_2.2.5
                 U __realpath_chk@@GLIBC_2.4
                 U round@@GLIBC_2.2.5
                 U _setjmp@@GLIBC_2.2.5
                 U sin@@GLIBC_2.2.5
                 U sinh@@GLIBC_2.2.5
                 U __snprintf_chk@@GLIBC_2.3.4
                 U __sprintf_chk@@GLIBC_2.3.4
                 U sqrt@@GLIBC_2.2.5
                 U __stack_chk_fail@@GLIBC_2.4
                 U __stpcpy_chk@@GLIBC_2.3.4
                 U strcat@@GLIBC_2.2.5
                 U strchr@@GLIBC_2.2.5
                 U strcmp@@GLIBC_2.2.5
                 U __strcpy_chk@@GLIBC_2.3.4
                 U strcpy@@GLIBC_2.2.5
                 U strerror@@GLIBC_2.2.5
                 U strlen@@GLIBC_2.2.5
                 U strncat@@GLIBC_2.2.5
                 U strncmp@@GLIBC_2.2.5
                 U strncpy@@GLIBC_2.2.5
                 U strrchr@@GLIBC_2.2.5
                 U strstr@@GLIBC_2.2.5
                 U strtol@@GLIBC_2.2.5
                 U tan@@GLIBC_2.2.5
                 U tanh@@GLIBC_2.2.5
                 U __vfprintf_chk@@GLIBC_2.3.4
                 U __vsnprintf_chk@@GLIBC_2.3.4
```

## strcpy

I don't see any guarantee that `strcpy` in `fxBuildModuleMap` is memory-safe. TODO: check whether `if (preparation && the->archive)` is always false in xsnap.

**Comment by erights** [comment]
Good! Yes, please move to Agoric.

On Wed, Jan 20, 2021 at 7:14 AM Dan Connolly <notifications@github.com>
wrote:

> @erights <https://github.com/erights> , you asked me how I'd go about
> this... the question has been rattling around in my head and it's starting
> to take shape... let me know if I should move this under the Agoric org.
> during this early brainstorming stage. ...

-- 
  Cheers,
  --MarkM

**Comment by dckc** [comment]
for reference, earlier discussions:
 - 2016: https://github.com/tc39/proposal-ses/issues/4 SES Proof of security
 - 2019: https://github.com/Agoric/agoric-sdk/issues/198 Chainmail info?

**Comment by dckc** [comment]
[splint][] looks somewhat promising... with a few annotations, it might be able to make a pretty good memory-safety argument.

The defaults seem incompatible with the current XS idioms, but perhaps that could be addressed easily:

```
moddable/xs/sources]$ splint +posixlib -I ../includes/ -I ../platforms/ xsDate.c
Splint 3.1.2 --- 02 Dec 2020

../platforms/xsPlatform.h:81:26: #error unknown compiler
  Preprocessing error. (Use -preproc to inhibit warning)
   In file included from xsCommon.h:41,
                 from xsAll.h:41,
                 from xsDate.c:38
xsAll.h:1942:85: Invalid character in macro parameter name
   In file included from xsDate.c:38
xsAll.h:1942:85: Parameter list for #define is not parseable
xsAll.h:1957:88: Invalid character in macro parameter name
...
*** Cannot continue.
```


[splint]: https://github.com/splintchecker/splint

found while looking for [CCured](http://people.eecs.berkeley.edu/~necula/Papers/ccured_toplas.pdf) via [SO clue](https://stackoverflow.com/questions/15230380/where-to-download-ccured).


## SAFECode


SAFECode: Enforcing Alias Analysis for Weakly Typed Languages
Dinakar Dhurjati, Sumant Kowshik, and Vikram Adve.
ACM SIGPLAN Conference on Programming Language Design and Implementation (PLDI), June 2006

https://news.ycombinator.com/item?id=17143237

[\[1705\.07354\] The Meaning of Memory Safety](https://arxiv.org/abs/1705.07354)

## Checked C looks active

[Checked C: Making C Safe by Extension \- Microsoft Research](https://www.microsoft.com/en-us/research/publication/checkedc-making-c-safe-by-extension/)
September 2018

https://github.com/Microsoft/checkedc 52d7a7f on Dec 18, 2020

https://github.com/Microsoft/checkedc-clang/releases 2020-07-31 2745f7d

**Comment by dckc** [comment]
### fuzzing

I played around with AFL tonight.
[AFL](https://github.com/google/AFL): american fuzzy lop - a security-oriented fuzzer

This looks like the result of some fuzzing too:

[SEGV at /xs/sources/xsBigInt\.c:1182 · Issue \#485 · Moddable\-OpenSource/moddable](https://github.com/Moddable-OpenSource/moddable/issues/485)

looking at other work by the reporter...
[kvenux/capflow](https://github.com/kvenux/capflow): Isabelle proof for CapFlow

and from there:
[A Verified Capability-Based Model for Information Flow Security With Dynamic Policies](https://ieeexplore.ieee.org/stamp/stamp.jsp?arnumber=8315441)

p.s. cf https://github.com/wasmerio/wasmer/tree/master/fuzz

**Comment by dckc** [comment]
## getters and setters vs. static analysis

> The main difference between SES and SES<sub>light</sub> is that SES supports getters/setters and SES<sub>light</sub> does not because they are not amenable to the analysis methods we considered practical and ...

 -- [Automated Analysis of Security-Critical JavaScript APIs](https://storage.googleapis.com/pub-tools-public-publication-data/pdf/37199.pdf)
found via [A Taxonomy of Security Issues \- Agoric](https://agoric.com/taxonomy-of-security-issues/)

**Comment by erights** [comment]
Yes. And JS has changed even more since that article was written.

**Comment by dckc** [comment]
to prevent escape from vat:
 - compile to WASM
 - seccomp

@erights wondered about in-vat mutual-suspicion boundaries:
 - zcf vs. the contract
 - liveSlots vs. running code
   - non-determinism
----
p.s. browsing...

[Formalizing the LLVM Intermediate Representation for Verified Program Transformations](https://repository.upenn.edu/cgi/viewcontent.cgi?article=1743&context=cis_papers)
[esbmc/esbmc: The efficient SMT\-based bounded model checker](https://github.com/esbmc/esbmc)
[static\-analysis\-engineering/CodeHawk\-C: CodeHawk C Analyzer: sound static analysis of memory safety \(undefined behavior\)](https://github.com/static-analysis-engineering/CodeHawk-C)

### setjmp / longjump

[Support for setjmp / longjmp · Issue \#2625 · rust\-lang/rfcs](https://github.com/rust-lang/rfcs/issues/2625)

frama-C doesn't support it.

**Comment by dckc** [comment]
### esmbc -- Efficient model checking with CI techniques

This looks like a particularly good match for XS:

[Continuous Verification of Large Embedded Software using SMT-Based Bounded Model Checking](https://ssvlab.github.io/lucasccordeiro/papers/ecbs2010.pdf)

The project is active; the maintainers responded to an issue within a day: https://github.com/esbmc/esbmc/issues/355#issuecomment-771552103

### Map2Check -- fuzzing and symbolic execution

esbmc contributors overlap with https://github.com/hbgit/Map2Check , which combines fuzzing and symbolic execution.

tried the release from late 2019; not promising:

```
~/opt/map2check$ ./map2check  --target-function fxBuildDate ~/projects/moddable/build/tmp/lin/debug/xst/xsDate.bc
Adopting z3 solver... 
Started Map2Check
std::bad_alloc
```

### nesCheck -- good performance on subset of C

adding "embedded" to our search terms turns up other interesting work...

[nesCheck](https://github.com/HexHive/nesCheck) is aimed at a constrained subset of C, but the results are interesting:
> Through extensive evaluation benchmarks – both on specifically constructed programs and the standard TinyOS applications – we showed the effectiveness of nesCheck in enforcing memory protection while minimizing the runtime performance overhead (0.84% on energy, 5.3% on code size, up to 8.4% on performance, and 16.7% on RAM).
-- [Midi et. a. 2017](https://hexhive.epfl.ch/publications/files/17AsiaCCS2.pdf)

**Comment by dckc** [comment]
I'm experimenting with coverity scan: https://scan.coverity.com/projects/dckc-moddable?tab=overview

hmm... more involve than my first impression suggested.

**Comment by dckc** [comment]
discovered https://github.com/google/oss-fuzz by way of https://twitter.com/LazyFishBarrel today

[fuzzing/libFuzzerTutorial\.md at master · google/fuzzing](https://github.com/google/fuzzing/blob/master/tutorial/libFuzzerTutorial.md)

[Google Online Security Blog: Improving open source security during the Google summer internship program](https://security.googleblog.com/2020/12/improving-open-source-security-during.html)

**Comment by dckc** [comment]
https://github.com/trailofbits/manticore , a Symbolic execution tool, can analyze WebAssembly . I wonder if analysis of XS compiled to WASM would be fruitful.

**Comment by dckc** [comment]
tangentially related... (i.e. above the xsnap layer... or is it?)

**Comment by dckc** [comment]
Another fuzzing expert is working on XS! Go @rain6851 !

https://github.com/Moddable-OpenSource/moddable/issues/580
thru
https://github.com/Moddable-OpenSource/moddable/issues/587

cc @erights

**Comment by dckc** [comment]
## safety, metering via WASM?

Crazy idea? Kill two birds with one stone by compiling xsnap to WASM:
 - metering - reuse WASM metering from other mature blockchain work ( https://github.com/CosmWasm/cosmwasm ? )
 - fail-stop safety: if XS isn't memory safe, WASM should make it fail-stop
   - caveat: is the WASM runtime memory safe? perhaps in safe rust? can we completely trust safety of rust?

cc @michaelfig @warner @erights @dtribble @kriskowal

**Comment by zmanian** [comment]
General information here
- There are two dominant standalone WASM execution ecosystems. The BytecodeAlliance ecosystem and Wasmer ecosystem. Cosmwasm and Near protocol user  Wasmer and Substrate uses the Bytecode Alliance ecosystem. 

- Both ecosystems have support for Go bindings and metering. Using wasmer might be slightly preferred because more reference code can be generated from cosmwasm.

- AFAIK, wasmer 1.0 massively reduced the amount of unsafe rust in their code base. Wasmer is also extensively fuzz tested to ensure the single pass JIT isn't a source of sandbox escapes.

**Comment by dtribble** [comment]
> ### safety, metering via WASM?
> Crazy idea? Kill two birds with one stone by compiling xsnap to WASM:
> * metering
> * fail-stop safety: if XS isn't memory safe, WASM should make it fail-stop

We considered this option (ty Zaki for the detailed options).  

WRT **metering**, that means that the actual metering costs now derive the the entire tool chain of compilation of C to WASM, and the presents as a mysterious mapping to user-specified code. It ensures that nothing escapes the metering, but it would also make it difficult to have select things with custom metering (e.g., if we wanted to charge fixed costs for some class of crypto operations to not penalize multisig).

WRT fail-stop **memory-safety**, we are already in position to have each vat run in a separate process, so WASM adds nothing there. It could conceivably provide stronger safety for multiple-vats within a single process, but for the cost of checking that taller tool chain, we could add other memory safety sanity checks.

I would like to see XS or something like it directly in Rust someday :).

**Comment by dckc** [comment]
> WRT **metering**, that means that the actual metering costs now derive the the entire tool chain of compilation of C to WASM,

Yes, I thought of that soon after I posted the idea. _Any_ change at the C layer would change metering results.
